lib/internal/source_map.js

```diff
@@ -0,0 +1,152 @@
+'use strict';
+
+// See https://sourcemaps.info/spec.html for SourceMap V3 specification.
+const { Buffer } = require('buffer');
+const debug = require('internal/util/debuglog').debuglog('source_map');
+const { dirname, resolve } = require('path');
+const fs = require('fs');
+const {
+  normalizeReferrerURL,
+} = require('internal/modules/cjs/helpers');
+const { JSON, Object } = primordials;
+// For cjs, since Module._cache is exposed to users, we use a WeakMap
+// keyed on module, facilitating garbage collection.
+const cjsSourceMapCache = new WeakMap();
+// The esm cache is not exposed to users, so we can use a Map keyed
+// on filenames.
+const esmSourceMapCache = new Map();
+const { fileURLToPath, URL } = require('url');
+
+function maybeCacheSourceMap(filename, content, cjsModuleInstance) {
+  if (!process.env.NODE_V8_COVERAGE) return;
+
+  let basePath;
+  try {
+    filename = normalizeReferrerURL(filename);
+    basePath = dirname(fileURLToPath(filename));
+  } catch (err) {
+    // This is most likely an [eval]-wrapper, which is currently not
+    // supported.
+    debug(err.stack);
+    return;
+  }
+
+  const match = content.match(/\/[*/]#\s+sourceMappingURL=(?<sourceMappingURL>[^\s]+)/);
+  if (match) {
+    if (cjsModuleInstance) {
+      cjsSourceMapCache.set(cjsModuleInstance, {
+        url: match.groups.sourceMappingURL,
+        data: dataFromUrl(basePath, match.groups.sourceMappingURL)
+      });
+    } else {
+      // If there is no cjsModuleInstance assume we are in a
+      // "modules/esm" context.
+      esmSourceMapCache.set(filename, {
+        url: match.groups.sourceMappingURL,
+        data: dataFromUrl(basePath, match.groups.sourceMappingURL)
+      });
+    }
+  }
+}
+
+function dataFromUrl(basePath, sourceMappingURL) {
+  try {
+    const url = new URL(sourceMappingURL);
+    switch (url.protocol) {
+      case 'data:':
+        return sourceMapFromDataUrl(basePath, url.pathname);
+      default:
+        debug(`unknown protocol ${url.protocol}`);
```

Comment:
As a future improvement we can support valid but not file URLs here (like http URLs) although I believe data urls are by far the most common with relative file URLs a close second